govcd/tm_content_library_item.go

```diff
@@ -0,0 +1,377 @@
+		config.ItemType = "TEMPLATE"
+	}
+	return createOuterEntity(cl.client, outerType, c, config)
+}
+
+// cleanupContentLibraryItemOnUploadError prevents leaving stranded Content Library Items when any step of the creation (upload)
+// process fails.
+func cleanupContentLibraryItemOnUploadError(client *Client, identifier string, originalError error) error {
+	var err error
+	var cli *ContentLibraryItem
+	if strings.Contains(identifier, "urn:vcloud:contentLibraryItem") {
+		cli, err = getContentLibraryItemById(client, identifier)
+	} else {
+		cli, err = getContentLibraryItemByName(client, identifier)
+	}
+	if ContainsNotFound(err) {
+		// Nothing to do
+		return originalError
+	}
+	if err != nil {
+		return fmt.Errorf("the Content Library Item creation failed with error: %s\nCleanup of stranded Content Library Item also failed: %s", originalError, err)
+	}
+	err = getContentLibraryItemUploadTask(cli, func(task *Task) error {
+		var innerErr error
+		if task == nil {
+			// The task does not exist, so we try to delete the Content Library Item directly
+			innerErr = cli.Delete()
+			if innerErr != nil {
+				return innerErr
+			}
+			return nil
+		}
+		// Task exists, so we cancel it and let TM do the cleanup
+		innerErr = task.CancelTask()
+		if innerErr != nil {
+			return innerErr
+		}
+		return nil
+	})
+	if err != nil {
+		return fmt.Errorf("the Content Library Item creation failed with error: %s\nCleanup of stranded Content Library Item also failed: %s", originalError, err)
+	}
+	return originalError
+}
+
+// uploadContentLibraryItemFile uploads a Content Library Item file from the given slice with the given file present on disk
+func uploadContentLibraryItemFile(name string, cli *ContentLibraryItem, filesToUpload []*types.ContentLibraryItemFile, localFilePath string) error {
+	if cli == nil || len(filesToUpload) == 0 {
+		return fmt.Errorf("the Content Library Item or its files cannot be nil / empty")
+	}
+
+	// We just want to upload the selected file (named after the 'name' input parameter)
+	var fileToUpload *types.ContentLibraryItemFile
+	for _, f := range filesToUpload {
+		if f.Name == name {
+			fileToUpload = f
+			break
+		}
+	}
+	if fileToUpload == nil {
+		return fmt.Errorf("'%s' not found among the Content Library Item '%s' files", name, cli.ContentLibraryItem.Name)
+	}
+	filesAbsPaths, tmpDir, err := util.Unpack(localFilePath)
+	if err != nil {
+		return fmt.Errorf("%s. Unpacked files for checking are accessible in: %s", err, tmpDir)
+	}
+	defer func() {
+		err = os.RemoveAll(tmpDir)
+		if err != nil {
+			util.Logger.Printf("[DEBUG] could not clean up tmp directory %s", tmpDir)
+		}
+	}()
+
+	ud := uploadDetails{
+		uploadLink:               strings.ReplaceAll(fileToUpload.TransferUrl, "/transfer", "/tm/transfer"), // TODO: TM: Workaround, the link is missing /tm in path, so it gives 404 as-is
+		uploadedBytes:            0,
+		fileSizeToUpload:         fileToUpload.ExpectedSizeBytes,
+		uploadPieceSize:          1024,
```

Comment:
Very good catch, +1. Was looking whether we don't miss out on anything compared to the previous implementation/interface, and it's this.

Reply:
Done. To prevent signature changes, instead of adding more arguments to the `CreateContentLibraryItem`, I created a new type `ContentLibraryItemUploadArguments` which can be used to set arguments.
This object now supports filePath and the piece upload size